modules/imgcodecs/src/grfmt_spng.cpp

```diff
@@ -694,14 +646,22 @@ void spngCvt_BGR2Gray_8u_C3C1R(const uchar *bgr, int bgr_step,
     int i;
     for (; size.height--; gray += gray_step)
     {
-        double cBGR0 = 0.1140441895;
-        double cBGR2 = 0.2989807129;
+        int cBGR0 = 3737;
+        int cBGR1 = 19234;
+        int cBGR2 = 9797;
+
         if (_swap_rb)
             std::swap(cBGR0, cBGR2);
         for (i = 0; i < size.width; i++, bgr += 3)
         {
-            int t = static_cast<int>(cBGR0 * bgr[0] + 0.5869750977 * bgr[1] + cBGR2 * bgr[2]);
-            gray[i] = (uchar)t;
+            if (bgr[0] != bgr[1] || bgr[0] != bgr[2])
```

Comment:
libpng code is like
```
                if (red != green || red != blue)
               {
                  rgb_error |= 1;
                  /* NOTE: this is the historical approach which simply
                   * truncates the results.
                   */
                  *(dp++) = (png_byte)((rc*red + gc*green + bc*blue)>>15);
               }

               else
                  *(dp++) = red;
```

let me check any performans difference with
`gray[i] = (uchar)((cBGR0 * bgr[0] + cBGR1 * bgr[1] + cBGR2 * bgr[2]) >> 15);`